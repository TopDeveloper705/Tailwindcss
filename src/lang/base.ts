export function isClass(obj: Record<string, unknown>): boolean {
  const isCtorClass = obj.constructor && obj.constructor.toString().substring(0, 5) === 'class';
  if(obj.prototype === undefined) return isCtorClass;
  const proptotype = obj.prototype as Record<string, unknown>;
  const isPrototypeCtorClass = proptotype.constructor && proptotype.constructor.toString && proptotype.constructor.toString().substring(0, 5) === 'class';
  return isCtorClass || isPrototypeCtorClass;
}

export function __call__(obj: Record<string, unknown>, args: unknown[]): unknown {
  // @ts-expect-error types are unnecessary
  return isClass(obj) ? new obj(...args) : obj(...args);
}
