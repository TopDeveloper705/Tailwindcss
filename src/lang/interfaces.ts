import type jsep from 'jsep';

export type operand = number | string | boolean;
export type unaryCallback = (a: operand) => operand;
export type binaryCallback = (a: operand, b: operand) => operand;

export type AnyExpression = jsep.ArrayExpression
  | jsep.BinaryExpression
  | jsep.MemberExpression
  | jsep.CallExpression
  | jsep.ConditionalExpression
  | jsep.Identifier
  | jsep.Literal
  | jsep.LogicalExpression
  | jsep.ThisExpression
  | jsep.UnaryExpression;
