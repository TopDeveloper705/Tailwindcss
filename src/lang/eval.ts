import type { MemberExpression, Expression, Identifier } from 'jsep';
import type { operand, unaryCallback, binaryCallback, AnyExpression } from './interfaces';

const toNumber = (n: operand) => typeof n === 'number' ? n : typeof n === 'string' ? parseFloat(n): n ? 1 : 0;

const binops: {[key:string]:binaryCallback} = {
  '||': (a, b) => a || b,
  '&&': (a, b) => a && b,

  '==': (a, b) => a === b,
  '!=': (a, b) => a !== b,
  '<': (a, b) => a < b,
  '>': (a, b) => a > b,
  '<=': (a, b) => a <= b,
  '>=': (a, b) => a >= b,

  '+': (a, b) => toNumber(a) + toNumber(b),
  '-': (a, b) => toNumber(a) - toNumber(b),
  '*': (a, b) => toNumber(a) * toNumber(b),
  '/': (a, b) => toNumber(a) / toNumber(b),
  '%': (a, b) => toNumber(a) % toNumber(b),
};

const unops: {[key:string]:unaryCallback} = {
  '-': a => -a,
  '+': a => +a,
  '!': a => !a,
};

function evaluateMember(node: MemberExpression, context: Record<string, unknown>) {
  const object = evaluate(node.object, context) as Record<string, unknown>;
  return node.computed ? [object, object[evaluate(node.property, context) as string]] : [object, object[(node.property as Identifier).name]];
}

export function evaluate(_node: Expression, context: Record<string, unknown>):unknown {

  const node = _node as AnyExpression;

  let caller = undefined;
  let fn = undefined;
  let assign = undefined;

  switch (node.type) {

  case 'ArrayExpression':
    return node.elements.map(i => evaluate(i, context));

  case 'BinaryExpression':
    return binops[node.operator](evaluate(node.left, context) as boolean, evaluate(node.right, context) as boolean);

  case 'CallExpression':
    if (node.callee.type === 'MemberExpression') {
      assign = evaluateMember(node.callee as MemberExpression, context);
      caller = assign[0];
      fn = assign[1];
    } else {
      fn = evaluate(node.callee, context);
    }
    if (typeof fn !== 'function') { return undefined; }
    return fn.apply(caller, node.arguments.map(i => evaluate(i, context)));

  case 'ConditionalExpression':
    return evaluate(node.test, context) ? evaluate(node.consequent, context) : evaluate(node.alternate, context);

  case 'Identifier':
    return context[node.name];

  case 'Literal':
    return node.value;

  case 'LogicalExpression':
    if (node.operator === 'or') return evaluate(node.left, context) || evaluate(node.right, context);
    if (node.operator === 'and') return evaluate(node.left, context) && evaluate(node.right, context);
    return binops[node.operator](evaluate(node.left, context) as boolean, evaluate(node.right, context) as boolean);

  case 'MemberExpression':
    return evaluateMember(node, context)[1];

  case 'ThisExpression':
    return context;

  case 'UnaryExpression':
    return unops[node.operator](evaluate(node.argument, context) as operand);

  default:
    return undefined;
  }
}
