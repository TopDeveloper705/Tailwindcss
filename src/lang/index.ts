export { default as Lexer } from './lexer';
export { default as Parser } from './parser';
export { default as Interpreter } from './interpreter';

export { evaluate as eval } from './eval';
