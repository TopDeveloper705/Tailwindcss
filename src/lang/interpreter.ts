import { Parser } from './parser';
import { TokenType, BinOp, UnaryOp, Num, Var, Assign, NoOp } from './tokens';
import type { Operand } from './tokens';

export default class Interpreter {
  parser: Parser;
  GLOBAL_SCOPE: {[key:string]: unknown} = {};

  constructor(parser: Parser) {
    this.parser = parser;
  }

  error(msg = 'Interpreter Error'): never {
    throw Error(msg);
  }

  visit(node: Operand): unknown {
    if (node instanceof Var) return this.visit_Var(node);
    if (node instanceof Assign) return this.visit_Assign(node);
    if (node instanceof Num) return this.visit_Num(node);
    if (node instanceof UnaryOp) return this.visit_UnaryOp(node);
    if (node instanceof BinOp) return this.visit_BinOp(node);
    if (node instanceof NoOp) return this.visit_NoOp();
    this.error();
  }

  visit_Num(node: Num): number {
    return node.value;
  }

  visit_BinOp(node: BinOp): number {
    const left_value = this.visit(node.left);
    const right_value = this.visit(node.right);
    if (typeof left_value === 'number' && typeof right_value === 'number') {
      switch (node.op.type) {
      case TokenType.PLUS:
        return left_value + right_value;
      case TokenType.MINUS:
        return left_value - right_value;
      case TokenType.MUL:
        return left_value * right_value;
      case TokenType.DIV:
        return left_value / right_value;
      }
    }
    this.error();
  }

  visit_UnaryOp(node: UnaryOp): number {
    const value = this.visit(node.expr);
    if (typeof value === 'number') {
      switch (node.op.type) {
      case TokenType.PLUS:
        return +value;
      case TokenType.MINUS:
        return -value;
      }
    }
    this.error();
  }

  visit_Assign(node: Assign): void {
    this.GLOBAL_SCOPE[node.left.value] = this.visit(node.right);
  }

  visit_Var(node: Var): unknown {
    const var_name = node.value;
    if (var_name in this.GLOBAL_SCOPE) {
      return this.GLOBAL_SCOPE[var_name];
    }
    this.error(`unknown variable '${var_name}'`);
  }

  visit_NoOp():void {
    return;
  }

  interpret(): unknown {
    return this.visit(this.parser.parse());
  }
}
