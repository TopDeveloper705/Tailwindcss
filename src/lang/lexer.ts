import { isSpace, isDigit, isAlpha } from './utils';
import { Token, TokenType, REVERSED_KEYWORDS } from './tokens';


export class Lexer {
  text: string;
  pos: number;
  current_char?: string;

  constructor(text: string) {
    this.text = text;
    this.pos = 0;
    this.current_char = this.text[this.pos];
  }

  error(msg = 'Invalid charater'): never {
    throw Error(msg);
  }

  advance(step = 1): void {
    this.pos += step;
    this.current_char = this.pos > this.text.length - 1 ? undefined : this.text[this.pos];
  }

  peek(): string | undefined {
    if (this.pos + 1 < this.text.length) {
      return this.text[this.pos + 1];
    }
  }

  skip_whitespace(): void {
    while (this.current_char !== undefined && isSpace(this.current_char)) {
      this.advance();
    }
  }

  string(char: '\'' | '"' | '`'): Token {
    let result = '';
    let prev = '';
    while(this.current_char !== undefined && (this.current_char !== char || prev === '\\')) {
      result += this.current_char;
      prev = this.current_char;
      this.advance();
    }
    if (result) this.advance(); // right quote
    if (char === '`') {
      return new Token(TokenType.TEMPLATE, result);
    }
    return new Token(TokenType.STRING, result);
  }

  keyword(): Token {
    // handle reversed keywords
    let result = '';
    while (this.current_char !== undefined && isAlpha(this.current_char)) {
      result += this.current_char;
      this.advance();
    }
    if (result in REVERSED_KEYWORDS) return REVERSED_KEYWORDS[result];
    this.error();
  }

  numeric(): Token {
    let result = '';
    while (this.current_char !== undefined && isDigit(this.current_char)) {
      // int
      result += this.current_char;
      this.advance();
    }
    if (this.current_char === '.') {
      // float
      result += '.';
      this.advance();
      while (this.current_char !== undefined && isDigit(this.current_char)) {
        result += this.current_char;
        this.advance();
      }
    }
    // size
    const next = this.text.slice(this.pos,);
    if (next.startsWith('rem')) {
      this.advance(3);
      return new Token(TokenType.REM, +result);
    }
    if (next.startsWith('px')) {
      this.advance(2);
      return new Token(TokenType.PIXEL, +result);
    }
    if (next.startsWith('em')) {
      this.advance(2);
      return new Token(TokenType.EM, +result);
    }
    if (next.startsWith('%')) {
      this.advance();
      return new Token(TokenType.PERCENTAGE, +result);
    }
    if (next.startsWith('s')) {
      this.advance();
      return new Token(TokenType.SECOND, +result);
    }
    if (next.startsWith('deg')) {
      this.advance(3);
      return new Token(TokenType.DEGREE, +result);
    }
    return new Token(TokenType.NUMBER, +result);
  }

  color(): Token {
    let result = '';
    while (this.current_char !== undefined && isAlpha(this.current_char)) {
      result += this.current_char;
      this.advance();
    }
    return new Token(TokenType.COLOR, '#' + result);
  }

  id(): Token {
    let result = '';
    while (this.current_char !== undefined && isAlpha(this.current_char)) {
      result += this.current_char;
      this.advance();
    }
    return new Token(TokenType.ID, result);
  }

  property(): Token {
    let result = '';
    let prev = '';
    while (this.current_char !== undefined && (this.current_char !== ';' || prev === '\\')) {
      result += this.current_char;
      prev = this.current_char;
      this.advance();
    }
    return new Token(/\${.*}/.test(result)? TokenType.TEMPLATE : TokenType.STRING, result);
  }

  unknown(): Token {
    let result = '';
    let prev = '';
    while (this.current_char !== undefined && (this.current_char !== '{' || prev === '\\')) {
      result += this.current_char;
      prev = this.current_char;
      this.advance();
    }
    return new Token(TokenType.ID, result.trimEnd());
  }

  peek_next_token(): Token {
    const pos = this.pos;
    const char = this.current_char;
    const token = this.get_next_token();
    this.pos = pos;
    this.current_char = char;
    return token;
  }

  get_next_token(): Token {
    while (this.current_char !== undefined) {
      if (isSpace(this.current_char)) {
        this.skip_whitespace();
        continue;
      }

      if (isDigit(this.current_char)) {
        return this.numeric();
      }

      if (isAlpha(this.current_char)) {
        return this.id();
      }

      switch (this.current_char) {
      case '@':
        this.advance();
        return this.keyword();
      case '$':
        this.advance();
        return new Token(TokenType.DOLLAR, '$');
      case '=':
        this.advance();
        return new Token(TokenType.ASSIGN, '=');
      case ':':
        this.advance();
        this.skip_whitespace();
        return this.property();
      case ';':
        this.advance();
        return new Token(TokenType.SEMI, ';');
      case '\'':
        this.advance();
        return this.string('\'');
      case '"':
        this.advance();
        return this.string('"');
      case '`':
        this.advance();
        return this.string('`');
      case '#':
        this.advance();
        return this.color();
      case '{':
        this.advance();
        return new Token(TokenType.LBRACKET, '{');
      case '}':
        this.advance();
        return new Token(TokenType.RBRACKET, '}');
      case '+':
        this.advance();
        return new Token(TokenType.PLUS, '+');
      case '-':
        this.advance();
        return new Token(TokenType.MINUS, '-');
      case '*':
        this.advance();
        return new Token(TokenType.MUL, '*');
      case '/':
        this.advance();
        return new Token(TokenType.DIV, '/');
      case '(':
        this.advance();
        return new Token(TokenType.LPAREN, '(');
      case ')':
        this.advance();
        return new Token(TokenType.RPAREN, ')');
      default:
        return this.unknown();
      }
    }
    return new Token(TokenType.EOF, undefined);
  }
}
