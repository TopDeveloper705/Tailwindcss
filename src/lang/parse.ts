import { isSpace, isDigit } from './utils';

export const TOKENS = {
  INTEGER: 'INTEGER',
  PLUS: 'PLUS',
  MINUS: 'MINUS',
  MUL: 'MUL',
  DIV: 'DIV',
  LPAREN: '(',
  RPAREN: ')',
  EOF: 'EOF',
};

export class Token {
  type: string;
  value?: string | number;

  constructor(type: string, value?: string | number) {
    this.type = type;
    this.value = value;
  }
}

export class Lexer {
  text: string;
  pos: number;
  current_char?: string;

  constructor(text: string) {
    this.text = text;
    this.pos = 0;
    this.current_char = this.text[this.pos];
  }

  error(): never {
    throw Error('Invalid charater');
  }

  advance(): void {
    this.pos ++;
    this.current_char = this.pos > this.text.length - 1 ? undefined : this.text[this.pos];
  }

  skip_whitespace(): void {
    while (this.current_char !== undefined && isSpace(this.current_char)) {
      this.advance();
    }
  }

  integer(): number {
    let result = '';
    while (this.current_char !== undefined && isDigit(this.current_char)) {
      result += this.current_char;
      this.advance();
    }
    return +result;
  }

  get_next_token(): Token {
    while (this.current_char !== undefined) {
      if (isSpace(this.current_char)) {
        this.skip_whitespace();
        continue;
      }

      if (isDigit(this.current_char)) {
        return new Token(TOKENS.INTEGER, this.integer());
      }

      switch (this.current_char) {
      case '+':
        this.advance();
        return new Token(TOKENS.PLUS, '+');
      case '-':
        this.advance();
        return new Token(TOKENS.MINUS, '+');
      case '*':
        this.advance();
        return new Token(TOKENS.MUL, '*');
      case '/':
        this.advance();
        return new Token(TOKENS.DIV, '/');
      case '(':
        this.advance();
        return new Token(TOKENS.LPAREN, '(');
      case ')':
        this.advance();
        return new Token(TOKENS.RPAREN, ')');
      }

    }
    return new Token(TOKENS.EOF, undefined);
  }

}

export class AST {
}

export class BinOp {
  left: Num | BinOp;
  op: Token;
  right: Num | BinOp;
  constructor(left: Num | BinOp, op: Token, right: Num | BinOp) {
    this.left = left;
    this.op = op;
    this.right = right;
  }
}

export class Num {
  token: Token;
  value: number;
  constructor(token: Token) {
    this.token = token;
    this.value = token.value as number;
  }
}

export class Parser {
  lexer: Lexer;
  current_token: Token;

  constructor(lexer: Lexer) {
    this.lexer = lexer;
    this.current_token = lexer.get_next_token();
  }

  error(): never {
    throw Error('Invalid syntax');
  }

  eat(token_type: string): void {
    if (this.current_token.type === token_type) {
      this.current_token = this.lexer.get_next_token();
    } else {
      this.error();
    }
  }

  factor(): Num | BinOp {
    // factor: INTEGER | LPAREN expr RPAREN
    let node;
    const token = this.current_token;
    switch (token.type) {
    case TOKENS.INTEGER:
      this.eat(TOKENS.INTEGER);
      return new Num(token);
    case TOKENS.LPAREN:
      this.eat(TOKENS.LPAREN);
      node = this.expr();
      this.eat(TOKENS.RPAREN);
      return node;
    default:
      this.error();
    }
  }

  term(): BinOp | Num {
    // term: factor ((MUL | DIV) factor)*
    let node = this.factor();
    while ([TOKENS.MUL, TOKENS.DIV].includes(this.current_token.type)) {
      const token = this.current_token;
      switch (token.type) {
      case TOKENS.MUL:
        this.eat(TOKENS.MUL);
        break;
      case TOKENS.DIV:
        this.eat(TOKENS.DIV);
        break;
      }
      node = new BinOp(node, token, this.factor());
    }
    return node;
  }

  expr(): Num | BinOp {
    // expr   : term ((PLUS | MINUS) term)*
    // term   : factor ((MUL | DIV) factor)*
    // factor : INTEGER | LPAREN expr RPAREN
    let node = this.term();
    while ([TOKENS.PLUS, TOKENS.MINUS].includes(this.current_token.type)) {
      const token = this.current_token;
      switch (token.type) {
      case TOKENS.PLUS:
        this.eat(TOKENS.PLUS);
        break;
      case TOKENS.MINUS:
        this.eat(TOKENS.MINUS);
        break;
      }
      node = new BinOp(node, token, this.term());
    }
    return node;
  }

  parse(): Num | BinOp {
    const result = this.expr();
    // recover state
    this.lexer = new Lexer(this.lexer.text);
    this.current_token = this.lexer.get_next_token();
    return result;
  }
}

export class Interpreter {
  parser: Parser;

  constructor(parser: Parser) {
    this.parser = parser;
  }

  error(): never {
    throw Error('interpret error');
  }

  visit(node: Num | BinOp): number {
    if (node instanceof Num) return this.visit_Num(node);
    if (node instanceof BinOp) return this.visit_BinOp(node);
    this.error();
  }

  visit_BinOp(node: BinOp): number {
    switch (node.op.type) {
    case TOKENS.PLUS:
      return this.visit(node.left) + this.visit(node.right);
    case TOKENS.MINUS:
      return this.visit(node.left) - this.visit(node.right);
    case TOKENS.MUL:
      return this.visit(node.left) * this.visit(node.right);
    case TOKENS.DIV:
      return this.visit(node.left) / this.visit(node.right);
    }
    this.error();
  }

  visit_Num(node: Num): number {
    return node.value;
  }

  interpret(): number {
    return this.visit(this.parser.parse());
  }
}
