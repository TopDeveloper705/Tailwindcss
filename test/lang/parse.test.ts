import { TOKENS, Token, Lexer, Parser, Interpreter } from '../../src/lang/parse';

describe('Windi Lang', () => {
  it('lexer', () => {
    const lexer = new Lexer('14 + 2 * 3 - 6 / 2');
    expect(lexer.get_next_token()).toEqual(new Token(TOKENS.INTEGER, 14));
    expect(lexer.get_next_token()).toEqual(new Token(TOKENS.PLUS, '+'));
    expect(lexer.get_next_token()).toEqual(new Token(TOKENS.INTEGER, 2));
    expect(lexer.get_next_token()).toEqual(new Token(TOKENS.MUL, '*'));
  });

  it('eval', () => {
    const lexer = new Lexer('7 + 3 * (10 / (12 / (3 + 1) - 1))');
    const parser = new Parser(lexer);
    expect(parser.parse()).toMatchSnapshot('ast');
    const interpreter = new Interpreter(parser);
    expect(interpreter.interpret()).toEqual(22);
  });
});
