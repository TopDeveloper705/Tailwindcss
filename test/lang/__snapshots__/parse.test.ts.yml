Tools / eval / ast / 0: |-
  {
    "left": {
      "token": {
        "type": "INTEGER",
        "value": 7
      },
      "value": 7
    },
    "op": {
      "type": "PLUS",
      "value": "+"
    },
    "right": {
      "left": {
        "token": {
          "type": "INTEGER",
          "value": 3
        },
        "value": 3
      },
      "op": {
        "type": "MUL",
        "value": "*"
      },
      "right": {
        "left": {
          "token": {
            "type": "INTEGER",
            "value": 10
          },
          "value": 10
        },
        "op": {
          "type": "DIV",
          "value": "/"
        },
        "right": {
          "left": {
            "left": {
              "token": {
                "type": "INTEGER",
                "value": 12
              },
              "value": 12
            },
            "op": {
              "type": "DIV",
              "value": "/"
            },
            "right": {
              "left": {
                "token": {
                  "type": "INTEGER",
                  "value": 3
                },
                "value": 3
              },
              "op": {
                "type": "PLUS",
                "value": "+"
              },
              "right": {
                "token": {
                  "type": "INTEGER",
                  "value": 1
                },
                "value": 1
              }
            }
          },
          "op": {
            "type": "MINUS",
            "value": "+"
          },
          "right": {
            "token": {
              "type": "INTEGER",
              "value": 1
            },
            "value": 1
          }
        }
      }
    }
  }
